File: runtime-agent/policy/stage1.go
package policy

import (
	"context"
	"regexp"
	"strings"
	"time"
)

// Stage1Engine runs the deterministic Stage 1 checks: regex / keyword /
// PII / tool firewall. Stateless and safe to share across goroutines.
type Stage1Engine struct{}

// NewStage1Engine returns a default Stage 1 engine.
func NewStage1Engine() *Stage1Engine { return &Stage1Engine{} }

// Evaluate runs the engine against one input against one policy.
// Returns a StageResult; never errors — Stage 1 always produces a
// verdict (matched / not-matched), and the orchestrator decides what
// the next step is based on policy.EnforcementLevel.
func (e *Stage1Engine) Evaluate(
	_ context.Context, in *Input, p *CompiledPolicy, environment string,
) StageResult {
	start := time.Now()

	// 1. Tool firewall first — applies when the input describes a tool call
	if in.ToolName != "" {
		if r, ok := e.checkToolFirewall(in.ToolName, p); ok {
			return stampLatency(r, start)
		}
	}

	// 2. Walk enabled rules. Block actions short-circuit; otherwise
	//    accumulate and pick the highest-severity match.
	var accumulated []StageResult
	for _, rule := range p.Rules {
		if !rule.Enabled {
			continue
		}
		if environment != "" && len(rule.Environments) > 0 && !contains(rule.Environments, environment) {
			continue
		}

		match, matched := e.matchRule(&rule, in.Text)
		if !matched {
			continue
		}
		if rule.Action == "block" {
			return stampLatency(match, start)
		}
		accumulated = append(accumulated, match)
	}

	if len(accumulated) > 0 {
		chosen := accumulated[0]
		for _, r := range accumulated[1:] {
			if SeverityRank(r.Severity) > SeverityRank(chosen.Severity) {
				chosen = r
			}
		}
		return stampLatency(chosen, start)
	}

	return stampLatency(StageResult{
		Stage:   ExitStage1Regex,
		Matched: false,
		Action:  ActionAllowed,
	}, start)
}

func (e *Stage1Engine) checkToolFirewall(
	toolName string, p *CompiledPolicy,
) (StageResult, bool) {
	if _, ok := p.ToolDenylist[toolName]; ok {
		return StageResult{
			Stage:      ExitStage1Regex,
			Matched:    true,
			Action:     ActionBlocked,
			Severity:   SeverityCritical,
			Category:   "unsafe_tool_use",
			RuleID:     "tool_firewall:denylist",
			Confidence: 1.0,
			Reason:     "tool on denylist: " + toolName,
			Evidence:   map[string]any{"tool_name": toolName},
		}, true
	}
	if _, ok := p.ToolApprovalRequired[toolName]; ok {
		return StageResult{
			Stage:      ExitStage1Regex,
			Matched:    true,
			Action:     ActionEscalated,
			Severity:   SeverityHigh,
			Category:   "unsafe_tool_use",
			RuleID:     "tool_firewall:approval_required",
			Confidence: 1.0,
			Reason:     "tool requires approval: " + toolName,
			Evidence:   map[string]any{"tool_name": toolName},
		}, true
	}
	// Allowlist enforcement: empty allowlist = no enforcement
	if len(p.ToolAllowlist) > 0 {
		if _, ok := p.ToolAllowlist[toolName]; !ok {
			return StageResult{
				Stage:      ExitStage1Regex,
				Matched:    true,
				Action:     ActionBlocked,
				Severity:   SeverityHigh,
				Category:   "unsafe_tool_use",
				RuleID:     "tool_firewall:not_allowlisted",
				Confidence: 1.0,
				Reason:     "tool not on allowlist: " + toolName,
				Evidence:   map[string]any{"tool_name": toolName},
			}, true
		}
	}
	return StageResult{}, false
}

func (e *Stage1Engine) matchRule(rule *CompiledRule, text string) (StageResult, bool) {
	switch rule.Type {
	case "regex":
		for _, pattern := range rule.RegexPatterns {
			if m := pattern.FindStringIndex(text); m != nil {
				return resultFor(rule, map[string]any{
					"matched_text": redactMatch(text[m[0]:m[1]]),
				}), true
			}
		}
	case "keyword":
		lowered := strings.ToLower(text)
		for _, kw := range rule.Keywords {
			if kw != "" && strings.Contains(lowered, kw) {
				return resultFor(rule, map[string]any{"keyword": kw}), true
			}
		}
	case "pii_pattern":
		for _, pattern := range rule.RegexPatterns {
			if m := pattern.FindString(text); m != "" {
				// Credit card regex over-fires; require Luhn
				if isCardPattern(pattern) && !LuhnCheck(m) {
					continue
				}
				return resultFor(rule, map[string]any{"pii_detected": true}), true
			}
		}
	}
	return StageResult{}, false
}

func resultFor(rule *CompiledRule, evidence map[string]any) StageResult {
	return StageResult{
		Stage:      ExitStage1Regex,
		Matched:    true,
		Action:     actionToTaken(rule.Action),
		Severity:   rule.Severity,
		Category:   rule.Category,
		RuleID:     fallback(rule.ID, rule.Name),
		Confidence: 1.0,
		Reason:     "matched " + rule.Type + " rule " + rule.Name,
		Evidence:   evidence,
	}
}

func actionToTaken(action string) Action {
	switch action {
	case "block":
		return ActionBlocked
	case "flag":
		return ActionFlagged
	case "modify":
		return ActionModified
	case "escalate":
		return ActionEscalated
	case "log_only":
		return ActionAllowed
	default:
		return ActionFlagged
	}
}

func redactMatch(s string) string {
	if len(s) <= 4 {
		return strings.Repeat("*", len(s))
	}
	return s[:2] + strings.Repeat("*", len(s)-4) + s[len(s)-2:]
}

func isCardPattern(p *regexp.Regexp) bool {
	// Sentinel match against the credit_card pattern.
	return strings.HasPrefix(p.String(), `\b(?:\d[- ]?){12,18}`)
}

func contains(haystack []string, needle string) bool {
	for _, v := range haystack {
		if v == needle {
			return true
		}
	}
	return false
}

func fallback(primary, secondary string) string {
	if primary != "" {
		return primary
	}
	return secondary
}

func stampLatency(r StageResult, start time.Time) StageResult {
	r.LatencyUS = time.Since(start).Microseconds()
	return r
}

File: backend/app/policy/stage1.py
"""Stage 1 policy enforcement — regex / keyword / PII / tool firewall.

The fast path. Sub-1ms p99 latency budget. No I/O, no allocations in
the hot loop, no external library calls. Every rule is pre-compiled
into a :class:`CompiledRule`; this module just walks the rule list and
runs match() against the input text.

Behavior summary:
- Walk all enabled rules in order. Stop at the first MATCH whose action
  is "block" — that's the most aggressive verdict and short-circuits.
- For non-blocking matches (flag / log_only), accumulate them but keep
  walking so we can find a more severe one if present.
- If no rule matches, return ``matched=False`` so the orchestrator can
  pass to Stage 2 (or just allow, in fast-mode).
- Tool firewall is checked separately when the input context includes
  a tool call.
"""

from __future__ import annotations

import re
import time
from typing import Any

from app.policy.compiled import CompiledPolicy, CompiledRule, luhn_check
from app.policy.types import (
    Direction,
    PolicyInput,
    Severity,
    StageResult,
)

# ─────────────────────────────────────────────── Engine


class Stage1RegexEngine:
    """Stage 1 deterministic engine. Stateless — safe to share across goroutines.

    Implementations of higher-level cache management (compiled policy
    lookup, environment filtering) live above this layer in the
    pipeline orchestrator.
    """

    async def evaluate(
        self,
        *,
        input_: PolicyInput,
        policy: CompiledPolicy,
        environment: str | None = None,
    ) -> StageResult:
        start = time.perf_counter_ns()

        # 1. Tool firewall — applies only when the input describes a tool call
        tool_call = input_.context.get("tool_call") if input_.context else None
        if tool_call:
            tool_result = self._check_tool_firewall(
                tool_name=str(tool_call.get("name", "")),
                policy=policy,
            )
            if tool_result is not None:
                return self._stamp_latency(tool_result, start)

        # 2. Walk enabled rules. Track the most severe match so we never miss
        #    a critical one because a low-severity rule fired first.
        accumulated: list[StageResult] = []
        for rule in policy.rules:
            if not rule.enabled:
                continue
            if environment and rule.environments and environment not in rule.environments:
                continue

            match = self._match_rule(rule=rule, text=input_.text)
            if match is None:
                continue

            # Block actions short-circuit immediately
            if rule.action == "block":
                return self._stamp_latency(match, start)
            accumulated.append(match)

        if accumulated:
            # Pick highest severity
            chosen = max(accumulated, key=lambda r: _SEVERITY_RANK[r.severity])
            return self._stamp_latency(chosen, start)

        # 3. No match
        return self._stamp_latency(
            StageResult(stage="stage1_regex", matched=False, action="allowed"),
            start,
        )

    # ─────────────────────────────────────────── helpers

    def _match_rule(self, *, rule: CompiledRule, text: str) -> StageResult | None:
        """Return a StageResult if the rule matches, else None."""
        if rule.type == "regex":
            for pattern in rule.regex_patterns:
                m = pattern.search(text)
                if m is not None:
                    return _result_for(rule=rule, evidence={"matched_text": _redact_match(m)})

        elif rule.type == "keyword":
            lowered = text.lower()
            for kw in rule.keywords:
                if kw and kw in lowered:
                    return _result_for(rule=rule, evidence={"keyword": kw})

        elif rule.type == "pii_pattern":
            for pattern in rule.regex_patterns:
                m = pattern.search(text)
                if m is None:
                    continue
                # Credit card regex over-fires; require Luhn
                if pattern.pattern.startswith(r"\b(?:\d[- ]?){12,18}"):
                    digits = re.sub(r"\D", "", m.group(0))
                    if not luhn_check(digits):
                        continue
                return _result_for(rule=rule, evidence={"pii_detected": True})

        # rate_limit and custom rule types are scaffolded but not implemented
        # in Sprint 2 — they require state (per-session counters) which
        # belongs in the orchestrator, not the stage engine. Skip.
        return None

    def _check_tool_firewall(self, *, tool_name: str, policy: CompiledPolicy) -> StageResult | None:
        if not tool_name:
            return None
        if tool_name in policy.tool_denylist:
            return StageResult(
                stage="stage1_regex",
                matched=True,
                action="blocked",
                severity="critical",
                category="unsafe_tool_use",
                rule_id="tool_firewall:denylist",
                confidence=1.0,
                reason=f"tool {tool_name!r} is on the denylist",
                evidence={"tool_name": tool_name},
            )
        if tool_name in policy.tool_approval_required:
            return StageResult(
                stage="stage1_regex",
                matched=True,
                action="escalated",
                severity="high",
                category="unsafe_tool_use",
                rule_id="tool_firewall:approval_required",
                confidence=1.0,
                reason=f"tool {tool_name!r} requires approval",
                evidence={"tool_name": tool_name},
            )
        # Allowlist enforcement: if an allowlist is configured and this tool
        # isn't on it, block. An empty allowlist is interpreted as "no
        # allowlist enforcement" rather than "block everything".
        if policy.tool_allowlist and tool_name not in policy.tool_allowlist:
            return StageResult(
                stage="stage1_regex",
                matched=True,
                action="blocked",
                severity="high",
                category="unsafe_tool_use",
                rule_id="tool_firewall:not_allowlisted",
                confidence=1.0,
                reason=f"tool {tool_name!r} is not on the allowlist",
                evidence={"tool_name": tool_name},
            )
        return None

    @staticmethod
    def _stamp_latency(result: StageResult, start_ns: int) -> StageResult:
        latency_us = (time.perf_counter_ns() - start_ns) // 1000
        # Single chokepoint for every Stage 1 result — stamp the honest mode.
        return StageResult(
            stage=result.stage,
            matched=result.matched,
            action=result.action,
            severity=result.severity,
            category=result.category,
            rule_id=result.rule_id,
            confidence=result.confidence,
            reason=result.reason,
            latency_us=int(latency_us),
            evidence=result.evidence,
            mode="stage1_regex",
        )


# ─────────────────────────────────────────────── helpers


_SEVERITY_RANK: dict[Severity, int] = {
    "info": 0,
    "low": 1,
    "medium": 2,
    "high": 3,
    "critical": 4,
}


def _result_for(*, rule: CompiledRule, evidence: dict[str, Any]) -> StageResult:
    """Build a StageResult from a matched rule. Confidence is always 1.0
    on Stage 1 matches because the rule was deterministic."""
    return StageResult(
        stage="stage1_regex",
        matched=True,
        action=_action_to_taken(rule.action),
        severity=rule.severity,
        category=rule.category,
        rule_id=rule.id or rule.name,
        confidence=1.0,
        reason=f"matched {rule.type} rule {rule.name!r}",
        evidence=evidence,
    )


_ACTION_MAP: dict[str, str] = {
    "block": "blocked",
    "flag": "flagged",
    "modify": "modified",
    "escalate": "escalated",
    "log_only": "allowed",
}


def _action_to_taken(action: str) -> Any:
    return _ACTION_MAP.get(action, "flagged")


def _redact_match(match: re.Match[str]) -> str:
    """Return a length-preserving redaction so audit logs don't echo PII."""
    raw = match.group(0)
    if len(raw) <= 4:
        return "*" * len(raw)
    # Keep the first and last 2 chars so an investigator can sanity-check
    # the regex was right, but never log the full sensitive value.
    return raw[:2] + "*" * (len(raw) - 4) + raw[-2:]


# ─────────────────────────────────────────────── Module-level convenience


_engine = Stage1RegexEngine()


async def evaluate(
    *,
    text: str,
    direction: Direction,
    policy: CompiledPolicy,
    asset_id: str | None = None,
    session_id: str | None = None,
    context: dict[str, Any] | None = None,
    environment: str | None = None,
) -> StageResult:
    """Top-level Stage 1 entrypoint.

    Constructs a PolicyInput from the loose arguments and invokes the
    shared engine. Convenience wrapper for callers that aren't going
    through the full pipeline orchestrator yet.
    """
    return await _engine.evaluate(
        input_=PolicyInput(
            text=text,
            direction=direction,
            asset_id=asset_id,
            session_id=session_id,
            context=context or {},
        ),
        policy=policy,
        environment=environment,
    )
